lua/entities/cfc_parachute/init.lua

```diff
@@ -0,0 +1,228 @@
+        if not IsValid( self ) then return end
+
+        self:Remove()
+    end )
+end
+
+function ENT:OnRemove()
+    timer.Remove( "CFC_Parachute_ExpireChute_" .. self:EntIndex() )
+    timer.Remove( "CFC_Parachute_FollowOwner_" .. self:EntIndex() )
+
+    local owner = self._chuteOwner
+    if not IsValid( owner ) then return end
+
+    owner.cfcParachuteChute = nil
+end
+
+function ENT:Think()
+    if not self._chuteIsOpen then return end
+
+    local owner = self._chuteOwner
+
+    if not IsValid( owner ) then
+        self:Remove()
+
+        return
+    end
+
+    self:SetAngles( owner:GetAngles() )
+    CFC_Parachute._ApplyChuteForces( owner, self )
+    self:NextThink( CurTime() )
+
+    return true
+end
+
+function ENT:CanOpen()
+    if self._chuteIsOpen then return false end
+
+    local owner = self._chuteOwner
+    if not IsValid( owner ) then return false end
+    if CFC_Parachute.IsPlayerCloseToGround( owner ) then return false end
+
+    return true
+end
+
+function ENT:ApplyLurch()
+    local lurchForce = -math.Rand( MIN_LURCH:GetFloat(), MAX_LURCH:GetFloat() )
+
+    self._chuteLurch = self._chuteLurch + lurchForce
+end
+
+function ENT:ApplyChuteDesign()
+    local owner = self._chuteOwner
+    if not IsValid( owner ) then return end
+
+    if not owner.cfcParachuteDesignID then
+        -- Requests the client to send their design selection since :GetInfoNum() is not behaving correctly even with FCVAR_USERINFO
+        -- Could be due to FCVAR_NEVER_AS_STRING if :GetInfoNum() expects a string that it then converts, without caring about the original type
+
+        net.Start( "CFC_Parachute_SelectDesign" )
+        net.Send( owner )
+
+        return
+    end
+
+    local designID = owner.cfcParachuteDesignID or 1
+    local materialName =
+        CFC_Parachute.DesignMaterialNames[designID] or
+        CFC_Parachute.DesignMaterialNames[1]
+    local fullMaterial = CFC_Parachute.DesignMaterialPrefix .. materialName
+
+    self:SetSubMaterial( 0, fullMaterial )
+end
+
+
+function ENT:_UpdateChuteDirection()
+    local chuteDirRel = Vector( self._chuteMoveForward - self._chuteMoveBack, self._chuteMoveRight - self._chuteMoveLeft, 0 )
+
+    net.Start( "CFC_Parachute_DefineChuteDir" )
+    net.WriteEntity( self )
+    net.WriteVector( chuteDirRel )
+    net.Broadcast()
+
+    self._chuteDirRel = chuteDirRel
+end
+
+function ENT:_KeyPress( ply, key, state )
+    if ply ~= self._chuteOwner then return end
+
+    if MOVE_KEY_LOOKUP[key] then
+        if key == IN_FORWARD then
+            self._chuteMoveForward = state and 1 or 0
+        elseif key == IN_BACK then
+            self._chuteMoveBack = state and 1 or 0
+        elseif key == IN_MOVERIGHT then
+            self._chuteMoveRight = state and 1 or 0
+        elseif key == IN_MOVELEFT then
+            self._chuteMoveLeft = state and 1 or 0
+        end
+
+        if not self._chuteIsOpen then return end
+
+        self:_UpdateChuteDirection()
+    end
+end
```

Comment:
Performance suggestion

```suggestion
do
    local IN_BACK = IN_BACK
    local IN_FORWARD = IN_FORWARD
    local IN_MOVELEFT  = IN_MOVELEFT
    local IN_MOVERIGHT = IN_MOVERIGHT

    function ENT:_KeyPress( ply, key, state )
        local selfTable = self:GetTable()
    
        if ply ~= selfTable._chuteOwner then return end
        if not MOVE_KEY_LOOKUP[key] then return end
    
        if key == IN_FORWARD then
            selfTable._chuteMoveForward = state and 1 or 0
        elseif key == IN_BACK then
            selfTable._chuteMoveBack = state and 1 or 0
        elseif key == IN_MOVERIGHT then
            selfTable._chuteMoveRight = state and 1 or 0
        elseif key == IN_MOVELEFT then
            selfTable._chuteMoveLeft = state and 1 or 0
        end
    
        if not selfTable._chuteIsOpen then return end
    
        self:_UpdateChuteDirection()
    end
end
```